examples/piecestore/rpc/client/main.go

```diff
@@ -0,0 +1,183 @@
+
+func main() {
+
+	app := cli.NewApp()
+
+	// Set up connection with rpc server
+	var conn *grpc.ClientConn
+	conn, err := grpc.Dial(":7777", grpc.WithInsecure())
+	if err != nil {
+		log.Fatalf("did not connect: %s", err)
+	}
+	defer conn.Close()
+	client := pb.NewPieceStoreRoutesClient(conn)
+	ctx := context.Background()
+
+	app.Commands = []cli.Command{
+		{
+			Name:    "upload",
+			Aliases: []string{"u"},
+			Usage:   "upload data",
+			Action: func(c *cli.Context) error {
+				if c.Args().Get(0) == "" {
+					return argError.New("No input file specified")
+				}
+
+				file, err := os.Open(c.Args().Get(0))
+				if err != nil {
+					return err
+				}
+				// Close the file when we are done
+				defer file.Close()
+
+				fileInfo, err := os.Stat(c.Args().Get(0))
+
+				if fileInfo.IsDir() {
+					return argError.New(fmt.Sprintf("Path (%s) is a directory, not a file", c.Args().Get(0)))
+				}
+
+				var fileOffset, storeOffset int64 = 0, 0
+				var length = fileInfo.Size()
+				var ttl = time.Now().Unix() + 86400
+
+				id, err := utils.DetermineID(file, fileOffset, length)
+				if err != nil {
+					return err
+				}
+
+				// Created a section reader so that we can concurrently retrieve the same file.
+				dataSection := io.NewSectionReader(file, fileOffset, length)
+
+				err = api.StorePieceRequest(ctx, client, id, dataSection, fileOffset, length, ttl, storeOffset)
+
+				if err != nil {
+					fmt.Printf("Failed to store file of id: %s\n", id)
+					return err
+				}
+
+				fmt.Printf("successfully Stored file of id: %s\n", id)
+				return nil
+			},
+		},
+		{
+			Name:    "download",
+			Aliases: []string{"d"},
+			Usage:   "download data",
+			Action: func(c *cli.Context) error {
+				if c.Args().Get(0) == "" {
+					return argError.New("No id specified")
+				}
+
+				id := c.Args().Get(0)
+
+				if c.Args().Get(1) == "" {
+					return argError.New("No output file specified")
+				}
+				_, err := os.Stat(c.Args().Get(1))
+
+				if !os.IsNotExist(err) {
+					return argError.New("Path already exists: " + c.Args().Get(1))
+				}
+
+				dataPath := c.Args().Get(1)
+
+				if err = os.MkdirAll(filepath.Dir(dataPath), 0700); err != nil {
+					return err
+				}
+
+				// Create File on file system
+				dataFile, err := os.OpenFile(dataPath, os.O_RDWR|os.O_CREATE, 0755)
+				if err != nil {
+					return err
+				}
+
+				pieceInfo, err := api.PieceMetaRequest(ctx, client, id)
+				if err != nil {
+					return err
+				}
+
+				reader, err := api.RetrievePieceRequest(ctx, client, id, 0, pieceInfo.Size)
+				if err != nil {
+					fmt.Printf("Failed to retrieve file of id: %s\n", id)
+					os.Remove(dataPath)
+					return err
+				}
+				defer reader.Close()
+
+				totalRead := int64(0)
+				for totalRead < pieceInfo.Size {
+					b := make([]byte, 4096)
+					n, err := reader.Read(b)
+					if err != nil {
+						if err == io.EOF {
+							break
+						}
+						return err
+					}
+
+					n, err = dataFile.Write(b[:n])
```

Comment:
io.Copy! you don't need to write all this

Reply:
if you want to only copy pieceInfo.Size, use io.LimitReader

Reply:
👍